src/backend/aspen/cli/sync_auth0.py

```diff
@@ -0,0 +1,420 @@
+        for obj in to_add:
+            create_callback(obj)
+        if self.delete_ok:
+            for obj in to_delete:
+                delete_callback(obj)
+
+    def sync_users(self):
+        db_users: MutableSequence[User] = (
+            self.db.execute(sa.select(User)).scalars().all()
+        )
+        auth0_users = self.auth0_client.get_users()
+        auth0_only, db_only = compare_stuff(
+            auth0_users, db_users, "user_id", "auth0_user_id"
+        )
+        user_manager = UserManager(self.auth0_client, self.db, self.dry_run)
+        self.update_objects(auth0_only, db_only, user_manager)
+
+    def sync_groups(self):
+        db_groups: MutableSequence[Group] = (
+            self.db.execute(sa.select(Group)).scalars().all()
+        )
+        auth0_orgs = self.auth0_client.get_orgs()
+        auth0_only, db_only = compare_stuff(
+            auth0_orgs, db_groups, "display_name", "name"
+        )
+        group_manager = GroupManager(self.auth0_client, self.db, self.dry_run)
+        self.update_objects(auth0_only, db_only, group_manager)
+
+    def sync_memberships(self):
+        auth0_orgs = self.auth0_client.get_orgs()
+        db_groups: MutableSequence[Group] = (
+            (self.db.execute(sa.select(Group))).scalars().all()
+        )
+        for org in auth0_orgs:
+            auth0_memberships = self.auth0_client.get_org_members(org)
+            # TODO, we might want to stuff the auth0 group ID's into the groups table to
+            # make this simpler.
+            db_group = [
+                group for group in db_groups if group.name == org["display_name"]
+            ]
+            if not db_group:
+                # We're assuming that at this point in the script, we've already sync'd
+                # whatever groups we plan to sync, and if we don't find matching groups
+                # in both data stores, it's intentional for some reason.
+                logging.warning(
+                    f"Skipping sync of group {org['display_name']} - no DB row"
+                )
+                continue
+            db_group = db_group[0]
+            db_group_users = (
+                self.db.execute(sa.select(User).where(User.group == db_group))
+                .scalars()
+                .all()
+            )
+            # TODO - right now if a user is a member of an auth0 org, we consider
+            # that "good enough" for the purposes of this sync script. We need to *ALSO*
+            # validate that they have the correct role, but that will be more important
+            # in the next version, so we're skipping that right now.
+            auth0_only, db_only = compare_stuff(
+                auth0_memberships, db_group_users, "user_id", "auth0_user_id"
+            )
+            group_manager = UserGroupManager(
+                self.auth0_client, self.db, self.dry_run, org, db_group
+            )
+            self.update_objects(auth0_only, db_only, group_manager)
```

Comment:
Be aware that when we first created the auth0 tenant, the sign-up flow on the user-facing auth0 login page was enabled (presumably by default). We had at least several dozen people sign up before we shut that off, so there are in fact many auth0 accounts present in our tenant that *are not actual users*. Right now I would consider the prod database to be the source of truth for that.

Reply:
> This looks good, I just have a few more general questions. This is just the first step towards an event-driven, user-initiated process for user management, right? Honestly it's unclear to me why this task is part of the view members ticket, and I'm unsure how this is going to help with the invitation flow.

For [reasons that I don't remember right now] we included "listing group invitations" in this ticket. We're planning to have Auth0 handle as much as possible for the user invitations flow, and in order to do that we need to have Auth0 groups for all of the groups in our DB. Some of the code in this script will be copied in to API endpoints for managing users/groups/etc as we build out this functionality. However, until we have our API managing all of our user/group objects, we're probably going to need to re-sync our db to Auth0 - for example, when we manually create a new db group. Even after we have full API coverage in place, having two sources of truth for the medium term makes it possible (or likely) that someone will manually update one system or another, and we'll need to re-sync these stores in either direction to make them consistent.

Reply:
> Be aware that when we first created the auth0 tenant, the sign-up flow on the user-facing auth0 login page was enabled (presumably by default). We had at least several dozen people sign up before we shut that off, so there are in fact many auth0 accounts present in our tenant that _are not actual users_. Right now I would consider the prod database to be the source of truth for that.

Yep, we'll probably run this script in a mode that lets it delete all Auth0 users that don't have matching DB rows in that case.